source/tools/testConsistency/dotnetWPF-renderer/WPFRenderer/WPFRenderer/Program.cs

```diff
@@ -0,0 +1,112 @@
+using AdaptiveCards.Templating;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace WPFRenderer
+{
+    public class Program
+    {
+        public static void Main()
+        {
+            JObject resultJson = new JObject();
+            List<string> errorList = new List<string>();
+            List<string> warningList = new List<string>();
+            resultJson.Add("imageData", "");
+            resultJson.Add("warnings", "");
+            resultJson.Add("errors", "");
+            Console.OutputEncoding = Encoding.UTF8;
+            try
+            {
+                StringBuilder inputSb = new StringBuilder();
+                string currLine;
+
+                //read a multi line input from stdin
+                do
+                {
+                    currLine = Console.ReadLine();
+                    inputSb = inputSb.Append(currLine);
+                }
+                while ((currLine != null) && (currLine != ""));
+
+                JObject jsonInput = JObject.Parse(inputSb.ToString());
+                string cardData = jsonInput.GetValue("data").ToString();
+                string cardTemplate = jsonInput.GetValue("template").ToString();
+
+                AdaptiveCardTemplate adaptiveCardTemplate = new AdaptiveCardTemplate(cardTemplate);
+                string adaptiveCardJson = adaptiveCardTemplate.Expand(cardData);
+                AdaptiveCardParseResult parseResult = AdaptiveCard.FromJson(adaptiveCardJson);
+                AdaptiveCardRenderer renderer = new AdaptiveCardRenderer();
+
+                GetResultJsonString(parseResult, renderer, resultJson, warningList).GetAwaiter().GetResult();
+            }
+
+            catch (Newtonsoft.Json.JsonReaderException e)
+            {
+                errorList.Add("Exception: " + e.Message);
+                errorList.Add("The input is not a valid json");
+            }
+            catch (AdaptiveTemplateException e)
+            {
+                errorList.Add("Exception: " + e.Message);
+                errorList.Add("Unable to bind the data with the template. The supplied card data is not valid");
+            }
+            catch (AdaptiveSerializationException e)
+            {
+                errorList.Add("Exception: " + e.Message);
+                errorList.Add("The supplied json input is not a valid Adaptive Card");
+
+            }
+            catch (ArgumentNullException e)
+            {
+                errorList.Add("Exception: " + e.Message);
+                errorList.Add("Unable to capture the image of the card. Ensure that the card template and card data are valid and try again.");
+            }
+            catch (Exception e)
+            {
+                errorList.Add("Exception: " + e.Message);
+                errorList.Add("An error was encountered: " + e.Message);
```

Comment:
How are you disambiguating WPF driver errors from say JS driver errors etc? Do we want to annotate the error/warning list or does it get disambiguated on the "client"?

Reply:
It gets disambiguated on the client. The drivers are not aware of other drivers and each sends its own set of results. These are all disambiguated at the client level. The client collects all the results and displays them appropriately